src/Workflow/StateGraph.php

```diff
@@ -0,0 +1,335 @@
+
+    /**
+     * Get the agent attached to a node.
+     */
+    public function getNode(string $name): AgentInterface
+    {
+        $this->assertNodeExists($name);
+
+        if ($name === self::START_NODE) {
+            throw new StateGraphError('START node is not attached to an agent');
+        }
+
+        if ($name === self::END_NODE) {
+            throw new StateGraphError('END node is not attached to an agent');
+        }
+
+        /** @var AgentInterface $node */
+        $node = $this->nodes[$name];
+
+        return $node;
+    }
+
+    /** @return string[] */
+    public function getSuccessors(string $node): array
+    {
+        $this->assertNodeExists($node);
+        return $this->edges[$node];
+    }
+
+    /** @return string[] */
+    public function getPredecessors(string $node): array
+    {
+        $this->assertNodeExists($node);
+
+        $predecessors = [];
+
+        foreach ($this->edges as $current => $successors) {
+            if (in_array($node, $successors)) {
+                $predecessors[] = $current;
+            }
+        }
+
+        return $predecessors;
+    }
+
+    public function nodeExists(string $name): bool
+    {
+        return array_key_exists($name, $this->nodes);
+    }
+
+    public function isStartConnectedToEnd(): bool
+    {
+        return $this->nodeIsConnectedTo(self::START_NODE, self::END_NODE);
+    }
+
+    /**
+     * Get a list of node names to evaluate to get from the start node to the end node.
+     *
+     * @return string[]
+     */
+    public function compile(): array
+    {
+        if (!$this->isStartConnectedToEnd()) {
+            throw new StateGraphError('Start node is not connected to end node');
+        }
+
+        if ($this->isCyclic()) {
+            throw new StateGraphError('The graph contains cycles');
+        }
+
+        /** @var array<string> */
+        $executionList = [];
+
+        /** @var array<string, bool> */
+        $evaluated = [];
+
+        foreach (array_keys($this->nodes) as $node) {
+            $evaluated[$node] = false;
+        }
+
+        $q = new \SplQueue();
+        $q->enqueue(self::START_NODE);
+        $evaluated[self::START_NODE] = true;
+
+        while (!$q->isEmpty()) {
+            /** @var string */
+            $node = $q->dequeue();
+
+            if ($node !== self::END_NODE) {
+                if ($this->pathExists($node, self::END_NODE, [$this, 'getSuccessors'])) {
+                    $this->evaluateNode($node, $evaluated, $executionList);
+
+                    foreach ($this->getSuccessors($node) as $successor) {
+                        if (!$evaluated[$successor]) {
+                            $q->enqueue($successor);
+                        }
+                    }
+                }
+            }
+        }
+
+        return $executionList;
+    }
+
+    /**
+     * Export the graph to Graphwiz format
+     * @see https://graphviz.org/doc/info/lang.html
+     */
+    public function toDot(): string
```

Comment:
It wouldn't cost much to have the almost-same function to export to Mermaid format.